### PR TITLE
KafkaV2SinkConnector

### PR DESCRIPTION
**Pull request by xinlian12**
Adding back Kafka sink connector change : https://github.com/Azure/azure-sdk-for-java/pull/38973

Extra changes:
- Added more shaded class files in guava.collect and guava.math
- Added the following two API in azure-cosmos
```
    PartitionKey toPartitionKey(List<Object> values, boolean strict);
    Mono<PartitionKeyDefinition> getPartitionKeyDefinition(CosmosAsyncContainer container);

```

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosWriteException.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+import org.apache.kafka.connect.errors.ConnectException;
+
+/**
+ * Generic CosmosDb sink write exceptions.
+ */
+public class CosmosWriteException extends ConnectException {
+    /**
+     *
+     */
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosWriteException.java:14]: empty comment box, we can remove it

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @xinlian12 
All my suggestions can be implemented in a follow up PR if it makes it easier.

File: sdk/cosmos/azure-cosmos-kafka-connect/pom.xml [hunk 2]
```diff
@@ -459,7 +492,7 @@ Licensed under the MIT License.
     </profile>
     <profile>
       <!-- integration tests, requires Cosmos DB Emulator Endpoint -->
-      <id>kafka-integration</id>
+      <id>kafka</id>
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/pom.xml:495]: will discuss offline for this

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/CosmosSinkConnector.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        for (int i = 0; i < maxTasks; i++) {
+            configs.add(this.sinkConfig.originalsStrings());
+        }
+
+        return configs;
+    }
+
+    @Override
+    public void stop() {
+        LOGGER.debug("Kafka Cosmos sink connector {} is stopped.");
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/CosmosSinkConnector.java:51]: info will be better, changed to use info

File: sdk/cosmos/azure-cosmos-kafka-connect/pom.xml [hunk 4]
```diff
@@ -47,11 +47,14 @@ Licensed under the MIT License.
       --add-opens com.azure.cosmos.kafka.connect/com.azure.cosmos.kafka.connect=ALL-UNNAMED
       --add-opens com.azure.cosmos.kafka.connect/com.azure.cosmos.kafka.connect.implementation=ALL-UNNAMED
       --add-opens com.azure.cosmos.kafka.connect/com.azure.cosmos.kafka.connect.implementation.source=com.fasterxml.jackson.databind,ALL-UNNAMED
+      --add-opens com.azure.cosmos.kafka.connect/com.azure.cosmos.kafka.connect.implementation.sink=ALL-UNNAMED
+      --add-opens com.azure.cosmos.kafka.connect/com.azure.cosmos.kafka.connect.implementation.sink.idStrategy=ALL-UNNAMED
+      --add-opens com.azure.cosmos/com.azure.cosmos.implementation=ALL-UNNAMED
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/pom.xml:52]: updated

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/CosmosSinkConnector.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A Sink connector that publishes topic messages to CosmosDB.
+ */
+public class CosmosSinkConnector extends SinkConnector {
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/CosmosSinkConnector.java:22]: Sink connector does not hold any resources like socket etc, so no real need to implement autocloseable

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/KafkaCosmosConfig.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+        return new ArrayList<>();
+    }
+
+    public static class AccountEndpointValidator implements ConfigDef.Validator {
+        @Override
+        @SuppressWarnings("unchecked")
+        public void ensureValid(String name, Object o) {
+            String accountEndpointUriString = (String) o;
+            if (StringUtils.isEmpty(accountEndpointUriString)) {
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/KafkaCosmosConfig.java:172]: StringUtils.isEmpty()  will cover null check as well

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/KafkaCosmosConstants.java [hunk 7]
```diff
@@ -3,13 +3,20 @@
 
 package com.azure.cosmos.kafka.connect.implementation;
 
+import com.azure.core.util.CoreUtils;
+
 public class KafkaCosmosConstants {
+    public static final String PROPERTIES_FILE_NAME = "azure-cosmos-kafka-connect.properties";
+    public static final String CURRENT_VERSION = CoreUtils.getProperties(PROPERTIES_FILE_NAME).get("version");
+    public static final String CURRENT_NAME = CoreUtils.getProperties(PROPERTIES_FILE_NAME).get("name");
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/KafkaCosmosConstants.java:11]: the name will be the artifactId, will discuss offline

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkTask.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+        if (this.sinkTaskConfig.getWriteConfig().isBulkEnabled()) {
+            this.cosmosWriter =
+                new KafkaCosmosBulkWriter(this.sinkTaskConfig.getWriteConfig(), this.context.errantRecordReporter());
+        } else {
+            this.cosmosWriter =
+                new KafkaCosmosPointWriter(this.sinkTaskConfig.getWriteConfig(), context.errantRecordReporter());
+        }
+
+        // TODO[public preview]: in V1, it will create the database if does not exists, but why?
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkTask.java:48]: there are few comments I left with // TODO[], will revisit all together once. But I want to remove this as I do not think it make sense to create the database here

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkTask.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    private IWriter cosmosWriter;
+
+    @Override
+    public String version() {
+        return KafkaCosmosConstants.CURRENT_VERSION;
+    }
+
+    @Override
+    public void start(Map<String, String> props) {
+        LOGGER.info("Starting the kafka cosmos sink task...");
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkTask.java:35]: updated

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkTask.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+            // transform sink records, for example populating id
+            List<SinkRecord> transformedRecords = sinkRecordTransformer.transform(containerName, entry.getValue());
+            this.cosmosWriter.write(container, transformedRecords);
+        }
+    }
+
+    @Override
+    public void stop() {
+        LOGGER.info("Stopping Kafka CosmosDB sink task...");
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkTask.java:88]: updated

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkTask.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        if (this.cosmosClient != null) {
+            this.cosmosClient.close();
+        }
+
+        this.cosmosClient = null;
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkTask.java:93]: can be removed.

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosWriteException.java (hunk 1 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosWriteException.java:14]: removed

**Pull request merged**